baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/MoreAbstractAsKeyOfSetOrMap.java

```diff
@@ -0,0 +1,139 @@
+            .onDescendantOf("com.google.common.cache.CacheBuilder")
+            .named("build");
+
+    private static final Matcher<ExpressionTree> CAFFEINE_CACHE_BUILDER = MethodMatchers.instanceMethod()
+            .onDescendantOf("com.github.benmanes.caffeine.cache.Caffeine")
+            .namedAnyOf("build", "buildAsync");
+
+    private static final Matcher<ExpressionTree> SET_COLLECTOR = MethodMatchers.staticMethod()
+            .onClass("java.util.stream.Collectors")
+            .named("toSet")
+            .withParameters();
+
+    private static final Matcher<ExpressionTree> UNMODIFIABLE_SET_COLLECTOR = MethodMatchers.staticMethod()
+            .onClass("java.util.stream.Collectors")
+            .named("toUnmodifiableSet")
+            .withParameters();
+
+    private static final Matcher<ExpressionTree> IMMUTABLE_SET_COLLECTOR = MethodMatchers.staticMethod()
+            .onClass("com.google.common.collect.ImmutableSet")
+            .named("toImmutableSet")
+            .withParameters();
+
+    private static final Matcher<MethodInvocationTree> MAP_COLLECTOR = Matchers.allOf(
+            // We could inspect the fourth argument for hash-based maps in the future. That method
+            // is relatively uncommon, and the other overloads use HashMap.
+            argumentCount(count -> count < 4),
+            MethodMatchers.staticMethod().onClass("java.util.stream.Collectors").named("toMap"));
+
+    private static final Matcher<MethodInvocationTree> CONCURRENT_MAP_COLLECTOR = Matchers.allOf(
+            // We could inspect the fourth argument for hash-based maps in the future. That method
+            // is relatively uncommon, and the other overloads use HashMap.
+            argumentCount(count -> count < 4),
+            MethodMatchers.staticMethod().onClass("java.util.stream.Collectors").named("toConcurrentMap"));
+
+    // Unlike map and concurrentMap collectors, 'toUnmodifiableMap' provides no overload to specify the map
+    // implementation.
+    private static final Matcher<ExpressionTree> UNMODIFIABLE_MAP_COLLECTOR =
+            MethodMatchers.staticMethod().onClass("java.util.stream.Collectors").named("toUnmodifiableMap");
+
+    private static final Matcher<ExpressionTree> IMMUTABLE_MAP_COLLECTOR = MethodMatchers.staticMethod()
+            .onClass("com.google.common.collect.ImmutableMap")
+            .named("toImmutableMap");
+
+    private static final Matcher<MethodInvocationTree> HASH_KEYED_METHODS =
+            Matchers.anyOf(GUAVA_CACHE_BUILDER, CAFFEINE_CACHE_BUILDER);
+
+    private static final Matcher<MethodInvocationTree> HASH_KEYED_COLLECTOR_METHODS = Matchers.anyOf(
+            SET_COLLECTOR,
+            UNMODIFIABLE_SET_COLLECTOR,
+            IMMUTABLE_SET_COLLECTOR,
+            MAP_COLLECTOR,
+            CONCURRENT_MAP_COLLECTOR,
+            UNMODIFIABLE_MAP_COLLECTOR,
+            IMMUTABLE_MAP_COLLECTOR);
```

Comment:
maybe overkill, but do we want to cover StreamEx `toImmutableMap` & `toImmutableSet` as well (`PreferImmutableStreamExCollections` pushes these when using `StreamEx`)?

Reply:
Let's punt those cases for now, I think they're less common than the stdlib/guava/cache cases, and I'm less familiar with StreamEx.

Reply:
Sounds good. We can also check the JDK 9+ `Map.of` & `Set.of` in a follow up .